ref-exchange/src/shares.rs

```diff
@@ -0,0 +1,122 @@
+use near_sdk::json_types::{ValidAccountId, U128};
+use near_sdk::{ext_contract, near_bindgen, PromiseOrValue};
+
+use crate::utils::{GAS_FOR_FT_TRANSFER_CALL, GAS_FOR_RESOLVE_TRANSFER, NO_DEPOSIT};
+use crate::*;
+
+#[ext_contract(ext_self)]
+trait ShareTokenResolver {
+    fn share_resolve_transfer(
+        &mut self,
+        pool_id: u64,
+        sender_id: AccountId,
+        receiver_id: AccountId,
+        amount: U128,
+    ) -> U128;
+}
+
+#[ext_contract(ext_share_token_receiver)]
+pub trait ShareTokenReceiver {
+    fn share_on_transfer(
+        &mut self,
+        pool_id: u64,
+        sender_id: AccountId,
+        amount: U128,
+        msg: String,
+    ) -> PromiseOrValue<U128>;
+}
+
+#[near_bindgen]
+impl Contract {
+    /// Transfer pool LP shares, if pool supports.
+    #[payable]
+    pub fn share_transfer(
+        &mut self,
+        pool_id: u64,
+        receiver_id: &ValidAccountId,
+        amount: U128,
+        memo: Option<String>,
+    ) {
+        assert_one_yocto();
+        let sender_id = env::predecessor_account_id();
+        let mut pool = self.pools.get(pool_id).expect("ERR_NO_POOL");
+        pool.share_transfer(receiver_id.as_ref(), amount.0);
+        self.pools.replace(pool_id, &pool);
+        log!(
+            "Transfer shares {} pool: {} from {} to {}",
+            pool_id,
+            amount.0,
+            sender_id,
+            receiver_id
+        );
+        if let Some(memo) = memo {
+            log!("Memo: {}", memo);
```

Comment:
let's add memo to the log above. What's the reason of doing 2 logs?

Reply:
Memo can be specifically parsed by various indexers

Reply:
It's still easy to do if it's at the end of the previous log. Addresses are well defined - easy for regexp.